design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+The testing tool also enables users to easily share their test files without any dependency on the execution environment. Therefore, it will be much easier to collaborate and design streaming pipelines with KSQL!
+
+
+## What is in scope
+
+The testing tool will be a self sufficient command line tool where users can pass JSON files describing test settings and configuration and evaluate the generated results.
+
+## What is not in scope
+
+N/A
+
+## Value/Return
+
+The proposed tool provides a new way of testing KSQL statements. It also is self sufficient and does not need any running environment such as Kafka or KSQL cluster.
+
+
+## Public APIS
+
+### Test File
+
+The test file is a JSON file that describes the test settings and test data. The following is an example test file:
+
+```JSON
+{
+ "comments": [
+   "Any comments on this test file!"
+ ],
+ "tests": [
+   {
+     "name": "Sample test",
+     "statements": [
+       "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+       "CREATE STREAM S1 as SELECT name FROM test where id > 100;"
+     ],
+     "inputs": [
+       {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
+       {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
+       {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
+     ],
+     "outputs": [
+       {"topic": "S1", "key": 101, "value": "101"}
+     ]
+   }
+ ]
+}
+```
+
+Each test file can include multiple tests. Each test can have the following attributes (these descriptions are the same as the ones described ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md))):
+
+- _name_ (required): The name of the test case.
+description (optional): A description of what the test case is testing.
+- _format_ (optional): An array of multiple different formats to run the test case as, e.g. AVRO, JSON, DELIMITED.
+- _statements_ (required): The list of statements to execute as this test case.
+- _properties_ (optional): A map of property name to value. Can contain any valid Ksql config. The config is passed to the engine when executing the statements in the test case
+- _topics_ (optional): An array of the topics this test case needs. Allows more information about the topic to be supplied, e.g. an existing Avro schema.
+- _inputs_ (required if `expectedException` not supplied): The set of input messages to be produced to Kafka topic(s).
+- _outputs_ (required if `expectedException` not supplied) The set of output messages expected in the output topic(s).
+- _expectedException_ (required if `inputs` and `outputs` not supplied) The exception that should be thrown when executing the supplied statements.
```

Comment:
Why would we expose `expectedException`?

Reply:
Mainly as we'll need it for our own testing.  We should be using the same tool internally. Well, not the command line variant, but the core logic.